astropy/io/fits/tests/test_image.py

```diff
@@ -73,7 +73,7 @@ def test_constructor_copies_header(self):
         # Original header should be unchanged
         assert phdr['FILENAME'] == 'labq01i3q_rawtag.fits'
 
-    @raises(ValueError)
+    @raises(IndexError)
```

Comment:
Just raising a normal `IndexError` from `list.__getitem__` might be surprising / unhelpful if the user was *expecting* there to be more HDUs (e.g. they know there are 5 HDUs and their codes does `hdulist[4]` after closing the file, and they are not aware of the lazy-loading behavior).

So perhaps as a compromise between removing backwards compatibility, but still being helpful to unwary users, `__getitem__` at the very least should catch this `IndexError` and re-raise it (I think `IndexError` is still appropriate) but with a more helpful error message.  Further, it should probably only do this if the *full file* hasn't been read before it was closed, which there might need to be a flag to track.  Currently this sets `self._read_all = True` in `_try_while_unread_hdus`, but this is a bit deceptive since it has not necessarily read to the end of the file.  So being more precise here might be helpful.